### ISSUE TITLE
[Proposal] Modeling

### ISSUE THREAD

**Issue by stealthycoin** [aws/chalice#1224]
Proof of concept: #1223 

Proposed interface
==============

Two additional keys on the route decorator. `input_model` and `output_model` (maybe request/response instead?) The value of these is either `None` (default) meaning no models are applied to this route. Or  an instance of `ModelConfig`. Which is defined as such

```python
class ModelConfig(object):
    def __init__(self, model, validate=False, cls=None):
        self.model = model
        self.validate = validate
        self.cls = cls
    ...
```

Where `model` is a marshmallow schema (currently class, though I plan to change it to instance), `validate` is whether or not the model should be enforced, and `cls` is a concrete data type that can be deserialized from the input body, or serialized to a response body.

The request object has a new property `model_body` which will be an instance of `cls` created from deserializing the `json_body`. If the `input_model` `cls` is not set then it will raise an error.

An instance of the `cls` on the `output_model` can be returned from a view function and will be automatically serialized using the schema.

Validate works slightly differently for in put and output. For input it will create a body validator in API gateway and enforce validation at the API Gateway layer. Invalid requests will not trigger a lambda and will return a 400.
For output API gateway does not provide any validation (understandably), so we provide it ourselves using the marshmallow schema. Validate will be manually called, and if any errors are discovered an error will be raised and the validation errors will be logged. A 500 will be returned to the user.

Sample Project
===========

```
$ tree .
.
├── app.py
├── chalicelib
│   ├── __init__.py
│   └── models.py
└── requirements.txt

1 directory, 4 files
```

app.py
```python
from chalice import Chalice
from chalice import ModelConfig

from chalicelib import models


app = Chalice(app_name='model-test')


@app.route('/hi', methods=['POST'],
           input_model=ModelConfig(
               models.UserSchema, validate=True, cls=models.User))
def say_hi():
    user = app.current_request.model_body
    return user.make_greeting()


@app.route('/get-user', methods=['POST'],
           output_model=ModelConfig(
               models.UserSchema, cls=models.User))
def get_user():
    totally_a_user_i_got_from_a_db = models.User(
        first_name='William',
        middle_name='Henry',
        last_name='Harrison',
    )
    return totally_a_user_i_got_from_a_db

```

chalicelib/models.py
```python
import random

from marshmallow import Schema, fields


class UserSchema(Schema):
    first_name = fields.String(required=True)
    middle_name = fields.String()
    last_name = fields.String(required=True)


class User(object):
    def __init__(self, first_name, last_name, middle_name=None):
        self.first_name = first_name
        self.middle_name = middle_name
        self.last_name = last_name

    def make_greeting(self):
        greeting = random.choice(['Hi', 'Hello', 'Sup'])
        if self.middle_name is None:
            return '%s %s %s' % (greeting, self.first_name, self.last_name)
        return '%s %s %s %s' % (
            greeting, self.first_name, self.middle_name, self.last_name)

```

Sample Calls
=========
first middle and last
```bash
$ echo '{"first_name": "george", "middle_name": "wilbur", "last_name": "francis"}' | http POST $(chalice url)hi
HTTP/1.1 200 OK

Hello george wilbur francis
```
first and last
```bash
$ echo '{"first_name": "george", "last_name": "francis"}' | http POST $(chalice url)hi
HTTP/1.1 200 OK

Hello george francis
```

just first
```bash
$ echo '{"first_name": "george"}' | http POST $(chalice url)hi
HTTP/1.1 400 Bad Request

{
    "message": "Invalid request body"
}
```

example showing output serialization of `User` object
```bash
$ http POST $(chalice url)get-user
HTTP/1.1 200 OK

{
    "first_name": "William",
    "last_name": "Harrison",
    "middle_name": "Henry"
}
```

**Comment by stealthycoin** [comment]
Another nice thing about buying into the marshmallow ecosystem is we can just depend on existing libs to solve serialization/deserialization. For example if you want to provide automatic serialization and deserialization you can rely on `marshmallow-dataclass` which lets us write code like this without any further changes to Chalice

```python
from chalice import Chalice, ModelConfig

from dataclasses import field
from marshmallow_dataclass import dataclass


app = Chalice(app_name='modelboy')
app.debug = True


@dataclass
class CreateUserInput(object):
    username: str = field(metadata={'required': True})
    password: str = field(metadata={'required': True})
    description: str = field(metadata={'required': True})


@dataclass
class CreateUserOutput(object):
    message: str = field(metadata={'required': True})

    @classmethod
    def from_username(cls, username):
        return cls(f"Successfully created username '{username}'")


@app.route('/', methods=['POST'],
          input_model=ModelConfig(CreateUserInput.Schema(), validate=True),
          output_model=ModelConfig(CreateUserOutput.Schema(), validate=True))
def index():
    data = app.current_request.json_body
    # data is an instance of CreateUserInput
    # write to database
    return CreateUserOutput.from_username(data.username)
```

By default if we were just using marshmallow we would get a validated dictionary back, by adding dataclasses we can get automatic object serialization/deserialization. I think this helps out enough with the default case of just providing a schema and getting validated JSON, that both should be included in a tutorial section.

The remaining issue is that the validation overlaps with the JSON schema validation that API gateway provides so it might be useful to separate it out into 2 flags in the `ModelConfig` initializer. This means there is a useless flag on the config if its used as an output validator since api gateway does not have response validation. It also clips off useful messages and returns a generic one.

ModelConfig(CreateUserInput.Schema(), validate=True, api_gateway_validate=True)
```
http POST $(chalice url) username=stealthycoin password=pass
HTTP/1.1 400 Bad Request
{
    "message": "Invalid request body"
}
```

ModelConfig(CreateUserInput.Schema(), validate=True, api_gateway_validate=False)
```
HTTP/1.1 400 Bad Request
{
    "Code": "BadRequestError",
    "Message": "BadRequestError: description: Missing data for required field."
}
```

The output validation can give useful information when debugging such as in this example where I return an empty string instead of the response object:
```
TypeError: {
  "message": [
    "Missing data for required field."
  ]
}
```

**Comment by dmulter** [comment]
Good stuff. Also might make sense to consider Swagger/OpenAPI model generation for input/output API models. Could implement this in a phase 2, but might be a good idea to consider in the design of a phase 1.

**Comment by ricky-sb** [comment]
I agree with you, we should just get rid of API Gateway validation and not deal with that dependency. Also, I would be more in favor of stacking decorators. Stuffing arguments into a single decorator can get a bit messy.    

What's the status on this @stealthycoin?

**Comment by jsfehler** [comment]
@stealthycoin Having had an immediate need for exposing documentation, I wrote a library that monkey patches chalice to add models. https://pypi.org/project/leangle/

It's a bit different from your implementation. It lacks validation and stacks decorators, but uses marshmallow in a similar way. Hope you find it interesting as a view on another way of getting similar results.

My biggest takeaway was that I'd rather not be monkey patching chalice if possible. If the SwaggerGenerator.generate_swagger exposed a way to hook in, I would be more protected from changes.

**Comment by saivarunk** [comment]
Any update on this proposal?

**Comment by skwashd** [comment]
Yes, please. This would simplify some of our existing logic. Marshmallow is great, but have you considered [pydantic](https://pypi.org/project/pydantic/)? For python 3.7+ it has no external dependencies, earlier versions depend on the dataclasses package.

**Comment by jakemwood** [comment]
Really looking forward to seeing progress on this issue, especially with [Goblet](https://github.com/goblet/goblet/blob/main/examples/main.py#L91) supporting modeling natively. If there has been an API agreed upon, I'd be happy to help contribute a solution.

**Comment by hbock-42** [comment]
Hello, any plan on working on this issue?
Being able to auto générate documentation natively would be very useful for nearly everyone.

**Comment by sagardevl** [comment]
Hi,
Can we hope this will be added soon.
